### PR TITLE
add deterministic k generation for ECC sign

### PR DESCRIPTION
**Pull request by JacobBarthelmeh**

### CODE REVIEW

**Comment by JacobBarthelmeh** [comment]
Adds the public API 

```
int wc_ecc_set_deterministic(ecc_key* key, byte flag);
int wc_ecc_gen_deterministic_k(const byte* hash, word32 hashSz,
        enum wc_HashType hashType, mp_int* priv, mp_int* k, mp_int* order,
        void* heap);
```

When the macro WOLFSSL_ECDSA_DETERMINISTIC_K is used. The deterministic generation of k follow RFC6967.

**Comment by toddouska** [comment]
@SparkiDev please review and then assign to me after approval.

File: wolfcrypt/src/ecc.c [hunk 1]
```diff
@@ -5299,6 +5301,48 @@ int wc_ecc_sign_hash(const byte* in, word32 inlen, byte* out, word32 *outlen,
...
+    DECLARE_CURVE_SPECS(curve, 1);
+    ALLOC_CURVE_SPECS(1);
+
+    /* get curve order */
+    err = wc_ecc_curve_load(key->dp, &curve, ECC_CURVE_FIELD_ORDER);
+    if (err == MP_OKAY) {
+        if (key->sign_k == NULL) {
+            key->sign_k = (mp_int*)XMALLOC(sizeof(mp_int), key->heap,
+                                                            DYNAMIC_TYPE_ECC);
+            if (key->sign_k) {
```
Comment (SparkiDev) [inline wolfcrypt/src/ecc.c:5318]: Only get here if sign_k was NULL?

File: wolfcrypt/src/ecc.c [hunk 2]
```diff
@@ -5299,6 +5301,48 @@ int wc_ecc_sign_hash(const byte* in, word32 inlen, byte* out, word32 *outlen,
...
+                if (mp_init(key->sign_k) != MP_OKAY ||
+                    wc_ecc_gen_deterministic_k(in, inlen,
+                            WC_HASH_TYPE_SHA256, &key->k, key->sign_k,
+                            curve->order, key->heap) != 0) {
+                    mp_free(key->sign_k);
+                    XFREE(key->sign_k, key->heap, DYNAMIC_TYPE_ECC);
+                    key->sign_k = NULL;
+                    wc_ecc_curve_free(curve);
+                    FREE_CURVE_SPECS();
+                    return ECC_PRIV_KEY_E;
```
Comment (SparkiDev) [inline wolfcrypt/src/ecc.c:5329]: Set err to ECC_PRIV_KEY_E and drop down to free below.

File: wolfcrypt/src/ecc.c [hunk 3]
```diff
@@ -5299,6 +5301,48 @@ int wc_ecc_sign_hash(const byte* in, word32 inlen, byte* out, word32 *outlen,
...
+                }
+            }
+            else {
+                wc_ecc_curve_free(curve);
```
Comment (SparkiDev) [inline wolfcrypt/src/ecc.c:5333]: Set err to MEMORY_E and drop down to free below.

File: wolfcrypt/src/ecc.c [hunk 4]
```diff
@@ -5756,6 +5821,264 @@ int wc_ecc_sign_hash_ex(const byte* in, word32 inlen, WC_RNG* rng,
...
+
+    if (ret == 0)
+        ret = wc_HmacFinal(&hmac, out);
+
+    wc_HmacFree(&hmac);
+    return ret;
+}
+
+
+/* Generates a deterministic key based of the message using RFC6967
```
Comment (SparkiDev) [inline wolfcrypt/src/ecc.c:5858]: Reference the RFC in function comment or in function itself.

File: wolfcrypt/src/ecc.c [hunk 5]
```diff
@@ -5756,6 +5821,264 @@ int wc_ecc_sign_hash_ex(const byte* in, word32 inlen, WC_RNG* rng,
...
+    }
+#endif
+
+    VSz = KSz = h1len = hashSz;
+    XMEMSET(V, 0x01, VSz);
+    XMEMSET(K, 0x00, KSz);
+
+    ret = mp_to_unsigned_bin(priv, x);
+
+    mp_init(&z1); /* always init z1 and free z1 */
```
Comment (SparkiDev) [inline wolfcrypt/src/ecc.c:5946]: move this to before mp_to_unsigned_bin.
(initialisation all together.)

File: wolfcrypt/src/ecc.c [hunk 6]
```diff
@@ -5756,6 +5821,264 @@ int wc_ecc_sign_hash_ex(const byte* in, word32 inlen, WC_RNG* rng,
...
+    if (ret == 0) {
+        qbits = mp_count_bits(order);
+    }
+
+    if (ret == 0) {
```
Comment (SparkiDev) [inline wolfcrypt/src/ecc.c:5951]: Remove this line as it's not needed.
Merge block with one above.

File: wolfcrypt/src/ecc.c [hunk 7]
```diff
@@ -5756,6 +5821,264 @@ int wc_ecc_sign_hash_ex(const byte* in, word32 inlen, WC_RNG* rng,
...
+        ret = mp_read_unsigned_bin(&z1, hash, hashSz);
+    }
+
+    if (ret == 0) {
+        /* right shift by bits in hash minus bits in order */
+        mp_rshb(&z1, (hashSz * WOLFSSL_BIT_SIZE) - qbits);
+        XMEMSET(h1, 0, sizeof(h1));
+    }
+
+    if (ret == 0) {
```
Comment (SparkiDev) [inline wolfcrypt/src/ecc.c:5961]: Merge block with one above.

File: wolfcrypt/src/ecc.c [hunk 8]
```diff
@@ -5756,6 +5821,264 @@ int wc_ecc_sign_hash_ex(const byte* in, word32 inlen, WC_RNG* rng,
...
+            }
+
+            if ((ret == 0) && ((int)(VSz * WOLFSSL_BIT_SIZE) != qbits)) {
+                /* handle odd case where shift of 'k' is needed */
+                mp_rshb(k, (VSz * WOLFSSL_BIT_SIZE) - qbits);
+            }
+
+            /* the key should be smaller than the order of base point */
+            if (ret == 0) {
+                if (mp_cmp(k, order) != MP_LT) {
```
Comment (SparkiDev) [inline wolfcrypt/src/ecc.c:6029]: pull up to previous line

File: wolfcrypt/src/ecc.c [hunk 9]
```diff
@@ -5756,6 +5821,264 @@ int wc_ecc_sign_hash_ex(const byte* in, word32 inlen, WC_RNG* rng,
...
+                    err = MP_VAL;
+                }
+            }
+
+            /* no 0 key's */
+            if (ret == 0) {
+                if (mp_iszero(k) == MP_YES)
```
Comment (SparkiDev) [inline wolfcrypt/src/ecc.c:6036]: pull up to previous line

File: wolfcrypt/src/ecc.c [hunk 10]
```diff
@@ -5756,6 +5821,264 @@ int wc_ecc_sign_hash_ex(const byte* in, word32 inlen, WC_RNG* rng,
...
+                    err = MP_ZERO_E;
+            }
+
+            /* if there was a problem with 'k' generated then try again */
+            if (ret == 0 && err != 0) {
+                ret = _HMAC_K(K, KSz, V, VSz, NULL, 0, NULL, 0, &intOct, K,
+                    hashType, heap);
+                if (ret == 0) {
+                ret = _HMAC_K(K, KSz, V, VSz, NULL, 0, NULL, 0, NULL, V,
```
Comment (SparkiDev) [inline wolfcrypt/src/ecc.c:6045]: formatting

File: wolfcrypt/src/ecc.c [hunk 11]
```diff
@@ -5756,6 +5821,264 @@ int wc_ecc_sign_hash_ex(const byte* in, word32 inlen, WC_RNG* rng,
...
+    }
+
+    if (ret == 0) {
+        x = (byte*)XMALLOC(MAX_ECC_BYTES, heap, DYNAMIC_TYPE_PRIVATE_KEY);
+        if (x == NULL)
+            ret = MEMORY_E;
+    }
+
+    /* bail out if any error has been hit at this point */
+    if (ret != 0) {
```
Comment (SparkiDev) [inline wolfcrypt/src/ecc.c:5927]: Could let it fall down to bottom where freeing is also done.

**Review by SparkiDev — CHANGES_REQUESTED** [review]

**Comment by SparkiDev** [comment]
Thanks for moving the deterministic generation into the sign operation.

File: wolfcrypt/src/ecc.c (hunk 8 above)
Comment (JacobBarthelmeh) [reply wolfcrypt/src/ecc.c:6029]: On this one am not sure about adding it to the previous statement block. The if check above it also has a check for the size of V value which is not needed with the compare.

File: wolfcrypt/src/ecc.c [hunk 12]
```diff
@@ -5756,6 +5821,266 @@ int wc_ecc_sign_hash_ex(const byte* in, word32 inlen, WC_RNG* rng,
...
+    if (ret == 0)
+        ret = wc_HmacUpdate(&hmac, x, xSz);
+
+    if (ret == 0)
+        ret = wc_HmacUpdate(&hmac, h1, h1Sz);
+
+    if (ret == 0)
+        ret = wc_HmacFinal(&hmac, out);
+
+    wc_HmacFree(&hmac);
```
Comment (SparkiDev) [inline wolfcrypt/src/ecc.c:5853]: wc_HmacInit() can fail and we don't want to call wc_HmacFree().

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: wolfcrypt/src/ecc.c [hunk 13]
```diff
@@ -5756,6 +5821,266 @@ int wc_ecc_sign_hash_ex(const byte* in, word32 inlen, WC_RNG* rng,
...
+
+            /* 3.2 step h.3 the key should be smaller than the order of base
+             * point */
+            if (ret == 0) {
+                if (mp_cmp(k, order) != MP_LT) {
+                    err = MP_VAL;
+                }
+
+                /* no 0 key's */
+                if (mp_iszero(k) == MP_YES)
```
Comment (SparkiDev) [inline wolfcrypt/src/ecc.c:6037]: May be:
  else if (...

**Review by SparkiDev — CHANGES_REQUESTED** [review]

**Review by SparkiDev — APPROVED** [review]

**Pull request merged**